keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,617 @@
+operations outside of Kubernetes, like creating or deleting volumes,
+or expanding the storage), it is expected that pod scheduling may
+still end up with a node from time to time where volume creation then
+fails. Rolling back in this case is complicated and outside of the
+scope of this KEP. For example, a pod might use two persistent
+volumes, of which one was created and the other not, and then it
+wouldn’t be obvious whether the existing volume can or should be
+deleted.
+
+For persistent volumes that get created independently of a pod nothing
+changes: it’s still the responsibility of the CSI driver to decide how
+to create the volume and then communicate back through topology
+information where pods using that volume need to run.
+
+Inline volumes could be extended to reference a storage class, which
+then could be used to handle more complex situations (like the LVM
+mirror vs. striped case) also for inline volumes. But this is outside
+the scope of this KEP.
+
+## Proposal
+
+### User Stories
+
+#### Ephemeral PMEM volume for Redis
+
+A [modified Redis server](https://github.com/pmem/redis) can use
+[PMEM](https://pmem.io/) via
+[memkind](http://memkind.github.io/memkind/) as DRAM replacement with
+higher capacity and lower cost at almost the same performance. When it
+starts, all old data is discarded, so an inline ephemeral volume is a
+suitable abstraction for declaring the need for a volume that is
+backed by PMEM and provided by
+[PMEM-CSI](https://github.com/intel/pmem-csi). But PMEM is a resource
+that is local to a node and thus the scheduler has to be aware whether
+enough of it is available on a node before assigning a pod to it.
+
+
+### Caching remaining capacity via the API server
+
+CSI defines the GetCapacity RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. The existing `CSIDriver` type gets
+extended with a `Status` field that (initially) only holds the
+capacity information in a format that is sufficiently flexible to
+handle the use cases outlined above. In the future, that `Status`
+field could also be extended to represent other dynamic information
+about the driver.
+
+The CSI driver deployment is already responsible for creating a
+`CSIDriver` object for itself. Now it will also have to enable and
+configure the adding, updating and removing of capacity
+information. To simplify that, the `external-provisioner` will be
+extended to handle this based on the already existing APIs (i.e. no
+changes needed in a CSI driver itself).
+
+The Kubernetes scheduler monitors `CSIDriver` objects and uses that
+information when it comes to making scheduling decisions. If the
+driver does not indicate that it supports capacity reporting, then the
+scheduler proceeds just as it does now, so nothing changes for
+existing CSI driver deployments.
+
+### API
+
+All Kubernetes API changes are under the `CSIStorageCapacity` feature
+gate.
```

Comment:
I chose `CSIStorageCapacity` at a time when this KEP had separate types for storage capacity. I kept it after moving to `CSIDriver.Status` because that name still describes the actual functionality.

However, `CSIDriverStatus` might be a better name now. I'm open for suggestions here...